web/apps/apps.go

```diff
@@ -3,21 +3,82 @@
 package apps
 
 import (
+	"html/template"
+	"io/ioutil"
 	"net/http"
 	"net/url"
 	"path"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/cozy/cozy-stack/pkg/apps"
 	"github.com/cozy/cozy-stack/pkg/couchdb"
+	"github.com/cozy/cozy-stack/pkg/crypto"
 	"github.com/cozy/cozy-stack/pkg/instance"
 	"github.com/cozy/cozy-stack/pkg/vfs"
 	"github.com/cozy/cozy-stack/web/jsonapi"
 	"github.com/cozy/cozy-stack/web/middlewares"
+	jwt "github.com/dgrijalva/jwt-go"
 	"github.com/labstack/echo"
 )
 
-const indexPage = "index.html"
+func buildCtxToken(i *instance.Instance, app *apps.Manifest, ctx apps.Context) string {
+	subject := "public"
+	if !ctx.Public {
+		subject = ctx.Folder
+	}
+	token, _ := crypto.NewJWT(i.SessionSecret, jwt.StandardClaims{
+		Audience: "context",
+		Issuer:   i.SubDomain(app.Slug),
+		IssuedAt: crypto.Timestamp(),
+		Subject:  subject,
+	})
+	return token
+}
+
+func serveApp(c echo.Context, i *instance.Instance, app *apps.Manifest, vpath string) error {
+	ctx, file := app.FindContext(vpath)
+	if ctx.NotFound() {
+		return echo.NewHTTPError(http.StatusNotFound, "Page not found")
+	}
+	if file == "" {
+		file = ctx.Index
+	}
+	filepath := path.Join(vfs.AppsDirName, app.Slug, ctx.Folder, file)
+	doc, err := vfs.GetFileDocFromPath(i, filepath)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusNotFound)
+	}
+	res := c.Response()
+	if file != ctx.Index {
+		return vfs.ServeFileContent(i, doc, "", c.Request(), res)
+	}
+
+	// For index file, we inject the stack domain and a context token
+	name, err := doc.Path(i)
+	if err != nil {
+		return err
+	}
+	content, err := i.FS().Open(name)
+	if err != nil {
+		return err
+	}
+	defer content.Close()
+	buf, err := ioutil.ReadAll(content)
+	if err != nil {
+		return err
+	}
+	tmpl, err := template.New(file).Parse(string(buf))
```

Comment:
Having to load the entire file just to see if it is a template is unfortunate. How could we improve this situation ?